### PR TITLE
Fixes #84 - Added BasicShaderUniformsTestSuite

### LINKED ISSUE #84: BasicShaderUniformsTestSuite: Create new TestSuite for testing gamemaker-specific built-in shader uniforms
### Description

A test suite should be developed and added to the framework to test the functionality of all gamemaker-specific built-in shader uniforms and constants mentioned on the [Shader Constants](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Asset_Management/Shaders/Shader_Constants.htm) manual page, as this is an area of functionality that is currently missing from the framework.

These tests were initially intended to be included in the BasicShaderTestSuite proposed in https://github.com/YoYoGames/GM-TestFramework/issues/59 but ended up taking up a sufficiently large amount of an already very large test suite that it made more sense to move them to their own test suite instead, in order to improve readability and granularity of tests.

### Requirements

The project should be up to date with 2024.6 release.

It should include tests that cover the following Gamemaker-specific built-in shader uniforms/constants:
- MATRIX_VIEW
- MATRIX_PROJECTION
- MATRIX_WORLD
- MATRIX_WORLD_VIEW
- MATRIX_WORLD_VIEW_PROJECTION
- MATRIX_MAX
- MAX_VS_LIGHTS
- gm_Matrices[matrix]
- gm_LightingEnabled
- gm_Lights_Direction[]
- gm_Lights_PosRange[]
- gm_Lights_Colour
- gm_AmbientColour
- gm_FogStart
- gm_RcpFogRange
- gm_PS_FogEnabled
- gm_FogColour
- gm_VS_FogEnabled
- gm_AlphaTestEnabled
- gm_AlphaRefValue

### Steps To The Task

- [x] Create a new test suite for testing built-in shader uniforms/constants
- [x] Add facts / theories to test the required uniforms/constants
- [x] Ensure tests are working correctly and in-line with documentation
- [x] Any errors discovered in GameMaker functionality or documentation during development should be reported as issues in the 
- [x] Any errors with the framework itself should be reported as issues within this repository

### PR DESCRIPTION
**Pull request by Indibulous**
Created BasicShaderUniformsTestSuite, a TestSuite containing 18 testAsyncs that test all the required functionality listed in https://github.com/YoYoGames/GM-TestFramework/issues/84:
- alpha_test
- gm_matrix_view
- gm_matrix_projection
- gm_matrix_world
- gm_matrix_world_view
- gm_matrix_world_view_projection
- matrix_max
- gm_vs_fog_enabled
- gm_ps_fog_enabled
- gm_fog_colour
- gm_fog_start
- gm_rcp_fog_range
- max_vs_lights
- gm_lighting_enabled
- gm_lights_direction
- gm_lights_pos_range
- gm_lights_colour
- gm_ambient_colour

Added 54 shaders in TestResources/shaders that are used in the tests:
 - 18 in GLSL
 - 18 in GLSLES
 - 18 in HLSL
 
Added a ShaderUniformTests folder in the "datafiles" folder within the project files, containing 70 expected buffer/image files used in the tests' draw buffer comparisons

### Currently Failing Tests

Using any target platform, there are 9 tests in this TestSuite that will fail due to known bugs - 6 due to [GameMaker-Bugs#6425](https://github.com/YoYoGames/GameMaker-Bugs/issues/6425):
 - alpha_test
 - gm_vs_fog_enabled
 - gm_ps_fog_enabled
 - gm_fog_colour
 - gm_fog_start
 - gm_rcp_fog_range

And 3 due to [GameMaker-Bugs#6509](https://github.com/YoYoGames/GameMaker-Bugs/issues/6509):
 - gm_lights_direction
 - gm_lights_pos_range
 - gm_lights_colour
 
These tests should hopefully begin passing once these issues are fixed.

There are additionally 4 tests that will fail when using a platform that makes use of the HLSL shader language (Windows or Xbox). This is due to the issue described in [GameMaker-Bugs#6840](https://github.com/YoYoGames/GameMaker-Bugs/issues/6840), where several shader uniforms will throw errors in HLSL and interrupt the entire test framework, so they have been temporarily commented out:
 - gm_lights_direction
 - gm_lights_pos_range
 - gm_lights_colour
 - gm_ambient_colour

These tests will need to be manually modified once [GameMaker-Bugs#6840](https://github.com/YoYoGames/GameMaker-Bugs/issues/6840) is fixed in order to un-comment the parts of code associated with the issue (which have been marked with comments), then they should begin passing as well.

### CODE REVIEW

File: projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml [hunk 1]
```diff
@@ -0,0 +1,1164 @@
...
+				var _info = os_get_info()
+				// If info map contains an entry with information about the platform's supported shader language..
+				if (ds_map_exists(_info, "GL_SHADING_LANGUAGE_VERSION"))
+				{
+					// If the shader language entry has a valid value (not undefined or empty string)
+					shader_language = _info[? "GL_SHADING_LANGUAGE_VERSION"];
+					if (!is_undefined(shader_language) && shader_language != "")
+					{
+						// Check that the shader language name doesn't have "GLSL ES" in it (because if it uses any version of GLSL ES then shaders should be supported)
+						assert_string_contains(shader_language, "OpenGL ES")
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml]:
```suggestion
						assert_string_contains(shader_language, "OpenGL ES");
```

File: projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml [hunk 2]
```diff
@@ -0,0 +1,1164 @@
...
+			test_shader = pick_shader_for_platform(sh_passthrough_glsles, sh_passthrough_hlsl, sh_passthrough_glsl);
+			// Check that the shader has been compiled
+			verify_shader_compiled(test_shader);
+			
+		},
+		ev_draw: function() {
+			// Start using shader
+			shader_set(test_shader);
+				// Test that shader_current() returns the shader we're currently using
+				assert_true(shader_current() == test_shader, test_current().name +", failed to get current shader")
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderTestSuite/BasicShaderTestSuite.gml]:
```suggestion
				assert_true(shader_current() == test_shader, test_current().name +", failed to get current shader");
```

File: projects/xUnit/scripts/BasicShaderUniformsTestSuite/BasicShaderUniformsTestSuite.gml [hunk 3]
```diff
@@ -0,0 +1,1198 @@
...
+				if (_set_value)
+				{
+					// If it should be set this frame, build a new matrix to set it as
+					_expected_matrix = [ 1, 0, 1, 0,
+										 0, 0, 0, 1,
+										 0, 0, 0, 1,
+										 1, 0, 1, 0 ]
+					matrix_set(matrix_world, _expected_matrix);
+					matrix_set(matrix_view, _expected_matrix);
+					_expected_matrix = matrix_multiply(matrix_get(matrix_world), matrix_get(matrix_view))
```
Comment (AdamBorek) [inline projects/xUnit/scripts/BasicShaderUniformsTestSuite/BasicShaderUniformsTestSuite.gml]:
```suggestion
					_expected_matrix = matrix_multiply(matrix_get(matrix_world), matrix_get(matrix_view));
```

**Review by AdamBorek — APPROVED** [review]

**Comment by DiasFranciscoA** [comment]
@Indibulous 
This looks fine is there any reason for all the other TestSuites to be commented out of execution though ? 🤔

**Comment by Indibulous** [comment]
Whoops! @DiasFranciscoA Nope, that's just me forgetting to uncomment them after testing 😅 should be fixed now.

**Review by DiasFranciscoA — APPROVED** [review]

**Pull request merged**